### ISSUE TITLE
Change failure result with config file

### ISSUE THREAD

**Issue by Lakitna** [byu-oit/openapi-enforcer#45]
I've been using `openapi-enforcer` for quite a while now for linting my OASs and it has been working great. So far I've been happy with what's considered an error and what a warning, until now. [_insert dramatic gopher_](https://giphy.com/gifs/kKdgdeuO2M08M)

I would like to change the warning you get when your OAS has an unknown format to an error. I looked in the docs and couldn't find anything about changing the result of a validation failure. Taking a look at the source tells me why it's not possible at the moment. https://github.com/byu-oit/openapi-enforcer/blob/master/src/enforcers/Parameter.js#L302

Can a configuration system like this be implemented in a way that does not require a full refactor? I would like to be able to set the result of a rule validation failure to `error`, `warning`, and `off`. Though I could do without the last one.

The reason I want to enforce format is to prevent typos from making it through the pipeline. Because the validation only gives a warning, the format `datee` will not fail the pipeline.

**Comment by Gi60s** [comment]
Nice gopher!

Are you concerned about the `format` for just the `Parameter` class or also for the `Schema` class? I'm not sure if that is really relevant to solving your current problem, just something to think about.

I like the idea of being able to tell specific warnings to produce errors instead and I don't think that would require a major refactor. It's probably something I could get started on next week. I'll leave this issue open until I can at least look into the work required there and get back to you.

Another alternative would be to use the [plugin functionality to create additional validation rules](https://byu-oit.github.io/openapi-enforcer/guide/component-plugins#add-custom-validation).

I haven't tested the below code, but something like this should work for your scenario until I get this feature implemented:

```js
const Enforcer = require('openapi-enforcer')

// 1. Get the component you want to customize
const Parameter = Enforcer.v2_0.Parameter

// 2. Call the extend function
Parameter.extend(function (data) {
    const { exception, enforcerData } = data
    if (this.type && this.format) {
        const enums = [];
        switch (parent.definition.type) {
            case 'file': enums.push('binary', 'byte'); break;
            case 'integer': enums.push('int32', 'int64'); break;
            case 'number': enums.push('float', 'double'); break;
            case 'string': enums.push('binary', 'byte', 'date', 'date-time', 'password'); break;
        }
        if (!enums.includes(format)) {
            exception.message('Non standard format used: ' + format);
        }
    }
})
```

**Comment by Lakitna** [comment]
Adding a rule would be a good workaround, but the code you provided would not cover custom formats. It's a good start though.

Is there anything I can do to help you with this?

**Comment by Gi60s** [comment]
FYI: work in progress.

**Comment by Gi60s** [comment]
Just wanted to give an update. I have been working on this. There are two ways to solve the problem, one is complicated but increases future flexibility and I'm leaning toward that. The other option is fairly easy but adds technical debt to the project.

Anyway, I wanted to let you know that this isn't forgotten and is actively being engineered.

**Comment by Lakitna** [comment]
Awesome, good to hear!

Is there anything I can do to help? Mind that I'll be on vacation until the 5th.

-------------

To me, the complicated way sounds like the way to go. I've implemented something similar before in [Rulebound](https://github.com/Lakitna/Rulebound). You might be able to get some inspiration there. My solution is a system where the user can add config on three levels:

1. Config from file or `package.json` (resolved using Cosmiconfig)

    Overwrites `config/defaults.ts`.

2. Config passed to the invocation function (same object structure as in 1)

    Overwrites result of 1.

3. Default config defined for the rule at its implementation

    Is overwritten by the result of 2.

Most of the implementation for this can be found in `src/config/manager.ts`.

**Comment by Gi60s** [comment]
I've opted to go with the easier method for implementing this for now. The alternative will require a major version change due to multiple breaking changes.

Changes are merged to master from PR #54.

Published to NPM as version `1.8.0`.

Documented as a component option `exceptionEscalateCodes` here: https://byu-oit.github.io/openapi-enforcer/api/openapi-enforcer

**Comment by Lakitna** [comment]
Awesome, that will work just fine for me :) Thank you for your effort!
